### PR TITLE
Support degenerate HostToBuffer cases

### PR DESCRIPTION
**Pull request by rsuderman**
It is possible for the transfer to be degenerate, specifically transferring either a splat matrix or a matrix with a length-0 dimension. This also includes a util for computing the permutation for when permuted examples appear.

### CODE REVIEW

File: iree/integrations/pjrt/common/api_impl.cc [hunk 1]
```diff
@@ -677,40 +678,55 @@ iree_status_t DeviceInstance::HostBufferToDevice(
...
+  bool is_dense_row_major = true;
+  for (int i = 0, s = num_dims; i < s; ++i) {
+    is_dense_row_major &= (shape[i] == dims[i]) && (perms[i] == i);
+    is_splat &= (byte_strides[i] == 0) || (dims[i] == 1);
+  }
+
+  if (!is_splat && !is_dense_row_major && !has_zero_length) {
+    return iree_make_status(
+        IREE_STATUS_UNIMPLEMENTED,
+        "only dense, row-major layouts currently supported");
```
Comment (jpienaar) [inline iree/integrations/pjrt/common/api_impl.cc:720]: Nit: Lets keep capitalization consistent between Status messages

File: iree/integrations/pjrt/common/tensor_utils.cc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  for (int i = 0; i < ndims; ++i) {
+    if (output_strides[i] == 0 || output_shape[i] == 1) {
+      reverse_perms[i] = i;
+    }
+  }
+
+  // Populate the reordered dimensions.
+  int idx = 0;
+  for (int i = 0; i < ndims; i++) {
+    if (reverse_perms[i] == -1) {
```
Comment (jpienaar) [inline iree/integrations/pjrt/common/tensor_utils.cc:44]: Is there a constant we can use for this?

**Review by jpienaar — APPROVED** [review]
Nice, could you also include a JAX unit test that exhibit or pass with this?

**Comment by stellaraccident** [comment]
We should have the cpp unit tests scripted, but in the mean time, can you just add a `bazel test iree/integrations/pjrt/common/...` to here: https://github.com/openxla/openxla-pjrt-plugin/blob/main/.github/workflows/build_smoketest.yml#L54

We'll want to do that better later but should get the coverage now.

File: iree/integrations/pjrt/common/tensor_utils.cc (hunk 2 above)
Comment (rsuderman) [reply iree/integrations/pjrt/common/tensor_utils.cc]: Declared a local constant. Best solution I could think of.

File: iree/integrations/pjrt/common/api_impl.cc (hunk 1 above)
Comment (rsuderman) [reply iree/integrations/pjrt/common/api_impl.cc:720]: Done (next one should not have been capitalized)

**Comment by rsuderman** [comment]
> We should have the cpp unit tests scripted, but in the mean time, can you just add a `bazel test iree/integrations/pjrt/common/...` to here: https://github.com/openxla/openxla-pjrt-plugin/blob/main/.github/workflows/build_smoketest.yml#L54
> 
> We'll want to do that better later but should get the coverage now.

Done.

**Comment by benvanik** [comment]
oh, you can use iree_hal_create_transfer_command_buffer to do fills as well, so you could avoid the if(splat) {begin/fill/end} and instead just use IREE_HAL_TRANSFER_COMMAND_TYPE_FILL instead of IREE_HAL_TRANSFER_COMMAND_TYPE_COPY.

**Pull request merged**